conformance/lang/expressions/type_cast_expression.balt

```diff
@@ -0,0 +1,1698 @@
+
+    record {|int a;string...; |} b = <record {|readonly int a;|}>{a: 5};
+    io:println(typeof b); // @output typedesc record {| readonly int a; |}
+
+    record {|int a;string...; |} c = <record {|readonly int a; "Car" b;|}>{a: 5, b: "Car"};
+    io:println(typeof c); // @output typedesc record {| readonly int a; "Car" b; |}
+
+    record {|int a;string...; |}|string d = <record {|readonly int a; "Car" b;|}|string>{a: 5, b: "Car"};
+    io:println(typeof d); // @output typedesc record {| readonly int a; "Car" b; |}
+
+    int[] e = <[int, byte]>[444, 21];
+    io:println(typeof e); // @output typedesc [int,byte]
+
+    int[]|byte[] f = <[int, byte]>[444, 21];
+    io:println(typeof f); // @output typedesc [int,byte]
+
+    any g = <float>1;
+    io:println(g); // @output 1.0
+
+    any h = <decimal>1.4;
+    io:println(h.toBalString()); // @output 1.4d
+}
+
+Test-Case: output
+Description: Test cast to readonly can be used to cause constructors withing the expression to construct values
+             with the read-only bit on.
+Labels: any, array-type, list-constructor-expr, map-type, mapping-constructor-expr, readonly-type, string, 
+        table-constructor-expr, table-type, type-cast-expr, typeof-expr
+
+function init() {
+    int[] a = <readonly>[1, 2];
+    io:println(typeof a); // @output typedesc [1,2]
+
+    any[] b = <readonly>[1, "Ballerina"];
+    io:println(typeof b); // @output typedesc [1,"Ballerina"]
+
+    map<string> c = <readonly>{a: "4", b: "car"};
+    io:println(typeof c); // @output typedesc {"a":"4","b":"car"}
+
+    map<any> d = <readonly>{a: 3, b: "car"};
+    io:println(typeof d); // @output typedesc {"a":3,"b":"car"}
+
+    table<record {|int a;|}> e = <readonly>table [
+            {a: 54}
+    ];
+    io:println(typeof e); // @output typedesc [{"a":54}]
+}
+
+Test-Case: output
+Description: Test cast to readonly can be used to verify that the value resulting from the evaluation of expression 
+             has its read-only bit on. 
+Labels: any, array-type, list-constructor-expr, map-type, mapping-constructor-expr, readonly-type, string, 
+        table-constructor-expr, table-type, type-cast-expr, typeof-expr
+
+function init() {
+    int[] a1 = <readonly>[1, 2];
+    int[] a = <readonly>a1;
+    io:println(typeof a); // @output typedesc [1,2]
+
+    any[] b1 = <readonly>[1, "Ballerina"];
+    any[] b = <readonly>b1;
+    io:println(typeof b); // @output typedesc [1,"Ballerina"]
+
+    map<string> c1 = <readonly>{a: "4", b: "car"};
+    map<string> c = <readonly>c1;
+    io:println(typeof c); // @output typedesc {"a":"4","b":"car"}
+
+    map<any> d1 = <readonly>{a: 3, b: "car"};
+    map<any> d = <readonly>d1;
+    io:println(typeof d); // @output typedesc {"a":3,"b":"car"}
+
+    table<record {|int a;|}> e1 = <readonly>table [
+            {a: 54}
+    ];
+    table<record {|int a;|}> e = <readonly>e1;
+    io:println(typeof e); // @output typedesc [{"a":54}]
+}
```

Comment:
Shall we add a test for this issue https://github.com/ballerina-platform/ballerina-lang/issues/24559